test/box-luatest/error_subsystem_improvements_test.lua

```diff
@@ -138,3 +140,51 @@ g.test_payload_inheritance = function()
     t.assert_equals(e3.bar, 33) -- Inherited from e2.
     t.assert_equals(e3.baz, 55) -- Own payload field.
 end
+
+-- Test the increased `box.error` serialization verbosity (gh-9105).
+g.test_increased_error_serialization_verbosity = function()
+    local prevprev = box.error.new{reason = "prevprev"}
+    local prev = box.error.new{reason = "prev"}
+    prev:set_prev(prevprev)
+    local cur = box.error.new{reason = "cur"}
+    cur:set_prev(prev)
+
+    t.assert_equals(cur:__serialize(), tostring(cur))
+    t.assert_equals(tostring(cur), cur.message)
+    t.assert_equals(json.encode(cur), json.encode(cur.message))
+
+    compat.box_error_serialize_verbose = 'new'
+
+    local cur_unpacked = cur:unpack()
+    local cur_serialized = cur:__serialize()
+
+    t.assert_equals(prevprev:unpack(), prevprev:__serialize())
+    t.assert_equals(type(cur_unpacked.prev), 'cdata')
+    t.assert_equals(type(cur_serialized.prev), 'table')
+
+    t.assert_equals(json.encode(cur), json.encode(cur_serialized))
+
+    local prevprev_unpacked = prevprev:unpack()
+    prevprev_unpacked.message = nil
+    prevprev_unpacked = json.encode(prevprev_unpacked)
+    t.assert_equals(tostring(prevprev),
+                    string.format("%s %s",
+                                  prevprev.message, prevprev_unpacked))
+
+    local prev_unpacked = prev:unpack()
+    prev_unpacked.message = nil
+    prev_unpacked.prev = nil
+    prev_unpacked = json.encode(prev_unpacked)
+    local cur_unpacked = cur:unpack()
+    cur_unpacked.message = nil
+    cur_unpacked.prev = nil
+    cur_unpacked = json.encode(cur_unpacked)
+    t.assert_equals(tostring(cur),
+                    string.format("%s %s\n" ..
+                                  "%s %s\n" ..
+                                  "%s %s",
+                                  prevprev.message, prevprev_unpacked,
+                                  prev.message, prev_unpacked,
+                                  cur.message, cur_unpacked))
+
+end
```

Comment:
- Let's restore `compat.box_error_serialize_verbose` to default in case of test failures.
- Let's split testing `tostring` and serializing
- I guess we need test serializing more thoroughly. We now only test that `prev` is just some table. 
- We don't need to test unpack I guess. 
- I would also test `yaml` serialization in the first place as it is what mentioned in the ticket and printed to console.

May be we'd better put this case into some existing diff test? We can check easily the encoder output and update it with `--update-result` in case of key order changes. AFAIK we plan to use a tweak to facilitate future change of `compat.box_error_serialize_verbose` to `new` for diff tests. Under this tweak we can filter filename and line info as discussed ealier.

Reply:
> Let's restore `compat.box_error_serialize_verbose` to default in case of test failures.

Done.

> Let's split testing `tostring` and serializing

Done.

> I guess we need test serializing more thoroughly. We now only test that `prev` is just some table.

No, we test that `__serialize` is equivalent to `unpack` (accurate to the `prev` field type), as per the functional requirements of #9105.

If `__serialize` is equivalent to `unpack`, then testing `serialize` is essentially testing `unpack`, which is already thoroughly tested, AFAIC.

> May be we'd better put this case into some existing diff test? We can check easily the encoder output and update it with --update-result in case of key order changes. AFAIK we plan to use a tweak to facilitate future change of compat.box_error_serialize_verbose to new for diff tests. Under this tweak we can filter filename and line info as discussed ealier.

I think this is redundant, see my comments above.

> We don't need to test `unpack` I guess.

I am not testing `unpack`. I am testing the equivalence of `__serialize` to `unpack` which is one of the essential requirements of #9105.

> I would also test `yaml` serialization in the first place as it is what mentioned in the ticket and printed to console.

Added `yaml` serialization test case.